### ISSUE TITLE
Certain videos fail to play

### ISSUE THREAD

**Issue by MegaBrutal** [wootguy/Radio#2]
I played on TWLZ Friday night (CET) and tried to play songs over the Radio. If few videos fail to play, I wouldn't be bothered, it's just part of life, but I actually had a hard time finding vids those finally played. I think the fail/play rate was like 50% which is kind of bad.  With this rate, I would assume it's not the individual videos causing the problem, but something could be probably improved in the Radio to have better compatibility with various videos.

Here is a lift of songs those didn't play (not all of them were tried by me):
https://www.youtube.com/watch?v=3f6EQAc3DQo
https://www.youtube.com/watch?v=AD5u1QtHu5I
https://www.youtube.com/watch?v=Am94HICuG7E
https://www.youtube.com/watch?v=BYavoagBysY
https://www.youtube.com/watch?v=oHME1-40pLI
https://www.youtube.com/watch?v=PgRHBk1K4ks

Honestly, I don't know what's wrong with these videos, though it would be nice if I could help to debug this.

I understand you're developing Radio as a favour and it's not my universal right to blast Assemblage 23 on TWLZ; so it's perfectly fine if you won't do anything about it. I'm already grateful for what we have now, I truly appreciate your efforts.

**Comment by wootguy** [comment]
it's something to do with how pafy/youtube-dl picks a link to download from. If I use the updated version of youtube-dl with pafy, then some videos fail to play, and the ones that do work will fail mid-stream every minute or so (`[Radio] playback failed at x. Attempting to restart.`). If using the older version, most videos play without problems, but age-restricted videos don't work. If I bypass pafy entirely, most videos work, but there's still the "playback failed at x" errors.

I think the python script should be using the older version of pafy for most links, and switch to the newer one only if the video is age-restricted.

**Comment by MegaBrutal** [comment]
That sounds like a workaround though. If the newer youtube-dl has problem, shouldn't it be reported to them as a regression?

**Comment by wootguy** [comment]
The age-restricted failure is reported, but the download failures could be a problem with youtube or ffmpeg. I don't think youtube-dl can predict which video stream format is going to prevent disconnects when streamed through ffmpeg. Maybe some query parameters are set wrong by youtube-dl. Maybe youtube kills the connection to fight against bots.

Sometimes the videos fail to start with a 403 error but you don't see that error in-game. Usually the video will work if you try it again a few seconds later.

**Comment by MegaBrutal** [comment]
Why is ffmpeg needed exactly? Does it convert/re-encode the audio to a format which can be fed into voice chat?

**Comment by wootguy** [comment]
Yeah, it converts to raw samples so that the steam_voice program doesn't need a bunch of different libraries to handle every video format. Ideally it wouldn't be used because it eats up a lot of memory when streaming multiple videos (about 3-10% per process).
